Rewrite ES6 syntax from idlharness.js into ES5

filmaj opened the pull request:
Relevant to the discussion in #6266 

Why? So that it can run on platforms that do not support ES6 yet. In my particular case, that is iOS 9.3. I think I will probably run into similar issues on older Android WebView / Browser implementations as well, although I have yet to do that.

Worth noting that at least the File API Blob Constructor tests might need a similar rewrite (or perhaps, conditional execution) to support the various `[Symbol.iterator]: value` object definition syntax.

With these changes I was able to successfully load and run the File API Manual IDL tests in a UIWebView instance on iOS 9.3. Without these changes, the webview would barf up syntax errors.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/6664)
<!-- Reviewable:end -->

filmaj committed:
#6266: rewrite ES6 syntax from idlharness into ES5, so that it can run on older platforms.

rwaldron commented on resources/idlharness.js:
```diff
@@ -415,14 +417,14 @@ IdlArray.prototype.is_json_type = function(type)
                            return mixin;
                        });
-                       if (mixins.some(m => m.has_to_json_regular_operation())) { return true; }
+                       if (mixins.some(function(m) { m.has_to_json_regular_operation() } )) { return true; }
```
The implicit return from the arrow needs to be an explicit `return ...` in the function expression.

tobie commented on resources/idlharness.js:
```diff
@@ -415,14 +417,14 @@ IdlArray.prototype.is_json_type = function(type)
                                throw new Error("Interface " + id + " not found (implemented by " + thing.name + ")");
                            }
                            return mixin;
                        });
```
I imagine you'll need to bind `this`, here, so: `mixins.map(function () { ... }, this)`

tobie reviewed (changes requested):
This looks lightweight enough indeed. Would be good to run the test suite before merging as it should have caught the missing binding of `this`.

tobie commented:
Note I have no idea if Travis runs the test suites for this. @jugglinmike?

filmaj commented:
My apologies. Let me fix up what shook out from the review so far and I'll make sure to familiarize myself with the tests and ensure those pass first.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259524290)

# Firefox (nightly) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259524291)

# Sauce (safari) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259524297)

# Chrome (unstable) #

tobie commented:
> I'll make sure to familiarize myself with the tests and ensure those pass first.

It's sort of really poorly documented at present I'm afraid. @jugglinmike, @jgraham should be able to help on #testing on W3C's irc. Happy to help too, but my python-fu is very mediocre.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259524298)

# Sauce (MicrosoftEdge) #

filmaj committed:
small tweaks re: implicit return and binding this while making idlharness.js ES5 compatible.

filmaj commented:
Hmm, the README and CONTRIBUTING docs only mention to run linting via `./wpt lint` (which passes). I've asked in IRC, see if anything shakes out there. If I can extract that information I will gladly add it to this PR to the README to help clear the path for future folks wanting to contribute :)

Thanks for the reviews thus far! I think I've addressed them.

tobie commented:
Yeah, so it happens that I was just about to start running some tests here again after a 4 week hiatus. Was planning to help document this too.

filmaj commented:
I have ran the `resources/test/` tests via `tox` and unfortunately I see a few failures (although they do seem unrelated to my changes to idlharness.js). The three failing tests are:

```
Example with iframe that consolidates tests via fetch_tests_from_window
Async Tests and Promises
Promise Tests
```

These are failing on both master as well as my branch.

The failing test output on master is here: https://gist.github.com/filmaj/51586a2d503847abad9077cc982d261b

I then checked out the SHA from my PR _prior_ to the feedback y'all gave + my subsequent changes (SHA 8615bfbe92) - the output is here: https://gist.github.com/filmaj/56348b337780fa9233a81f8fb47c59cf

That yielded a _new_ failure: `::IdlArray.prototype.is_json_type() <- . FAILED`

Finally, I checked out the latest commit from my PR, addressing your feedback, and that seems to at least pass the IdlArray failure. Output: https://gist.github.com/filmaj/10f1b0b5015a3dc3db4d1942c35cecca

A little bit of a manual CI run above, so sorry for the verbosity, but I think that at least gives me some confidence that I didn't introduce any regressions in idlharness.js with these changes.
